{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n",
      "true\n"
     ]
    }
   ],
   "source": [
    "#First of all, I do a function to open the file and order the data in a list\n",
    "def first_column(file, lista):\n",
    "    gfile = open(file, \"r\")\n",
    "    n = 0\n",
    "    for line in gfile:\n",
    "        if n == 0:\n",
    "            n += 1\n",
    "        else:\n",
    "            line = line.split('\\t')[0]\n",
    "            lista.append(line)\n",
    "#Second, I do a function to compare the list of germplasm with the list of locus\n",
    "def comp_list(lista1, lista2):\n",
    "    n = 0\n",
    "    for x in lista1:\n",
    "        if x == lista2[n]:\n",
    "            n += 1\n",
    "            print(\"true\")\n",
    "        else:\n",
    "            print(lista1[n] + \"is not here\")\n",
    "#Finally, I create both lists and do the functions\n",
    "A = []\n",
    "first_column(\"Germplasm.tsv\", A)\n",
    "B = []\n",
    "first_column(\"LocusGene.tsv\", B)\n",
    "\n",
    "comp_list(A, B)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Connected: root@mysql'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext sql\n",
    "#%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "(pymysql.err.ProgrammingError) (1007, \"Can't create database 'exam2'; database exists\")\n",
      "[SQL: create database exam2]\n",
      "(Background on this error at: http://sqlalche.me/e/f405)\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "5 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>information_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>exam2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>mysql</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>performance_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>sys</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('information_schema',),\n",
       " ('exam2',),\n",
       " ('mysql',),\n",
       " ('performance_schema',),\n",
       " ('sys',)]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Creating a database\n",
    "%sql create database exam2\n",
    "%sql show databases\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Tables_in_exam2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasm</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('germplasm',)]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#use the database and show tables\n",
    "%sql use exam2\n",
    "%sql show tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "5 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Field</th>\n",
       "        <th>Type</th>\n",
       "        <th>Null</th>\n",
       "        <th>Key</th>\n",
       "        <th>Default</th>\n",
       "        <th>Extra</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>id</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td>PRI</td>\n",
       "        <td>None</td>\n",
       "        <td>auto_increment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Locus</td>\n",
       "        <td>varchar(20)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasm</td>\n",
       "        <td>varchar(30)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>phenotype</td>\n",
       "        <td>varchar(500)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>pubmed</td>\n",
       "        <td>varchar(400)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('id', 'int(11)', 'NO', 'PRI', None, 'auto_increment'),\n",
       " ('Locus', 'varchar(20)', 'NO', '', None, ''),\n",
       " ('germplasm', 'varchar(30)', 'NO', '', None, ''),\n",
       " ('phenotype', 'varchar(500)', 'NO', '', None, ''),\n",
       " ('pubmed', 'varchar(400)', 'NO', '', None, '')]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#creating a table named germplasm\n",
    "%sql CREATE TABLE germplasm(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, Locus VARCHAR(20) NOT NULL, germplasm VARCHAR(30) NOT NULL, phenotype VARCHAR(500) NOT NULL, pubmed VARCHAR(400) NOT NULL);\n",
    "%sql DESCRIBE germplasm\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "4 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Field</th>\n",
       "        <th>Type</th>\n",
       "        <th>Null</th>\n",
       "        <th>Key</th>\n",
       "        <th>Default</th>\n",
       "        <th>Extra</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>id</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td>PRI</td>\n",
       "        <td>None</td>\n",
       "        <td>auto_increment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Locus</td>\n",
       "        <td>varchar(20)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Gene</td>\n",
       "        <td>varchar(30)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ProteinLength</td>\n",
       "        <td>float</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('id', 'int(11)', 'NO', 'PRI', None, 'auto_increment'),\n",
       " ('Locus', 'varchar(20)', 'NO', '', None, ''),\n",
       " ('Gene', 'varchar(30)', 'NO', '', None, ''),\n",
       " ('ProteinLength', 'float', 'NO', '', None, '')]"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#creating a table named locus\n",
    "%sql CREATE TABLE locus(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, Locus VARCHAR(20) NOT NULL, Gene VARCHAR(30) NOT NULL, ProteinLength FLOAT NOT NULL);\n",
    "%sql DESCRIBE locus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OrderedDict([('Locus', 'AT1G01040'), ('germplasm', 'CS3828'), ('phenotype', 'Increased abundance of miRNA precursors.'), ('pubmed', '17369351')])\n",
      "OrderedDict([('Locus', 'AT1G01060'), ('germplasm', 'lhy-101'), ('phenotype', 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.'), ('pubmed', '16891401')])\n",
      "OrderedDict([('Locus', 'AT1G01140'), ('germplasm', 'SALK_058629'), ('phenotype', 'hypersensitive to low potassium media'), ('pubmed', '17486125')])\n",
      "OrderedDict([('Locus', 'AT1G01220'), ('germplasm', 'SALK_012400C'), ('phenotype', 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.'), ('pubmed', '18199744')])\n",
      "OrderedDict([('Locus', 'AT2G03720'), ('germplasm', 'SALK_042433'), ('phenotype', 'Multiple straight hairs'), ('pubmed', '16367956')])\n",
      "OrderedDict([('Locus', 'AT2G03800'), ('germplasm', 'gek1-1'), ('phenotype', 'Ethanol hypersensitivity.'), ('pubmed', '15215505')])\n",
      "OrderedDict([('Locus', 'AT2G04240'), ('germplasm', 'xerico'), ('phenotype', 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.'), ('pubmed', '17933900')])\n",
      "OrderedDict([('Locus', 'AT2G05210'), ('germplasm', 'pot1-1'), ('phenotype', 'No visible phenotype.'), ('pubmed', '17627276')])\n",
      "OrderedDict([('Locus', 'AT3G02130'), ('germplasm', 'rpk2-2'), ('phenotype', 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.'), ('pubmed', '17419837')])\n",
      "OrderedDict([('Locus', 'AT3G02140'), ('germplasm', 'afp4-1'), ('phenotype', 'Decreased germination on high concentrations of glucose and sorbitol.'), ('pubmed', '18484180')])\n",
      "OrderedDict([('Locus', 'AT3G02230'), ('germplasm', 'rgp1-1'), ('phenotype', 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.'), ('pubmed', '21478444')])\n",
      "OrderedDict([('Locus', 'AT3G02260'), ('germplasm', 'tir3-1 RGLG1:rglg1 rglg2'), ('phenotype', 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.'), ('pubmed', '17586653')])\n",
      "OrderedDict([('Locus', 'AT3G02310'), ('germplasm', 'sep2-1'), ('phenotype', 'Non-described subtle phenotype.'), ('pubmed', '10821278')])\n",
      "OrderedDict([('Locus', 'AT3G02680'), ('germplasm', 'atnbs1-1'), ('phenotype', 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.'), ('pubmed', '17672843')])\n",
      "OrderedDict([('Locus', 'AT3G02850'), ('germplasm', 'CS3816'), ('phenotype', 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.'), ('pubmed', '17568770')])\n",
      "OrderedDict([('Locus', 'AT3G02870'), ('germplasm', 'vtc4-1'), ('phenotype', 'ascorbate deficient'), ('pubmed', '16595667')])\n",
      "OrderedDict([('Locus', 'AT3G03260'), ('germplasm', 'hdg8-1'), ('phenotype', 'No visible phenotype.'), ('pubmed', '16778018')])\n",
      "OrderedDict([('Locus', 'AT4G14790'), ('germplasm', 'pdd17'), ('phenotype', 'Defective pollen development.'), ('pubmed', '19237690')])\n",
      "OrderedDict([('Locus', 'AT4G15210'), ('germplasm', 'bmy1-2'), ('phenotype', 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.'), ('pubmed', '16297066')])\n",
      "OrderedDict([('Locus', 'AT4G15560'), ('germplasm', 'cla1-1'), ('phenotype', 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.'), ('pubmed', '10982425')])\n",
      "OrderedDict([('Locus', 'AT4G15570'), ('germplasm', 'maa3'), ('phenotype', 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.'), ('pubmed', '18772186')])\n",
      "OrderedDict([('Locus', 'AT4G15802'), ('germplasm', 'Athspb-2'), ('phenotype', 'Early flowering, reduced fertility, aborted seeds.'), ('pubmed', '20388662')])\n",
      "OrderedDict([('Locus', 'AT4G15880'), ('germplasm', 'esd4-2'), ('phenotype', 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.'), ('pubmed', '17513499')])\n",
      "OrderedDict([('Locus', 'AT4G16420'), ('germplasm', 'prz1-1'), ('phenotype', 'Altered response to auxin and cytokinin'), ('pubmed', '12747832')])\n",
      "OrderedDict([('Locus', 'AT4G16480'), ('germplasm', 'atint4-2'), ('phenotype', 'No visible phenotype.'), ('pubmed', '16603666')])\n",
      "OrderedDict([('Locus', 'AT5G10480'), ('germplasm', 'pas2-3'), ('phenotype', 'Segregates 25% embryo lethal.'), ('pubmed', '18799749')])\n",
      "OrderedDict([('Locus', 'AT5G10510'), ('germplasm', 'plt3-1'), ('phenotype', 'Short roots and shortened root meristem.'), ('pubmed', '17960244')])\n",
      "OrderedDict([('Locus', 'AT5G11110'), ('germplasm', 'kns2'), ('phenotype', 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.'), ('pubmed', '18779216')])\n",
      "OrderedDict([('Locus', 'AT5G11260'), ('germplasm', 'hy5-101'), ('phenotype', 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.'), ('pubmed', '16891401')])\n",
      "OrderedDict([('Locus', 'AT5G11510'), ('germplasm', 'myb3r4-1'), ('phenotype', 'No visible phenotype.'), ('pubmed', '17287251')])\n",
      "OrderedDict([('Locus', 'AT5G12200'), ('germplasm', 'pyd2-2'), ('phenotype', 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.'), ('pubmed', '19413687')])\n",
      "OrderedDict([('Locus', 'AT5G13290'), ('germplasm', 'crn-1'), ('phenotype', 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.'), ('pubmed', '12345678')])\n"
     ]
    }
   ],
   "source": [
    "#insert the data in the table of database\n",
    "import pymysql.cursors\n",
    "import csv\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                            user='root',\n",
    "                            password='root',\n",
    "                            db='exam2',\n",
    "                            charset='utf8mb4', \n",
    "                            cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    " \n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open('Germplasm.tsv') as csvfile:\n",
    "            germplasm = csv.DictReader(csvfile, delimiter = '\\t')\n",
    "            for row in germplasm:\n",
    "                print(row)\n",
    "                sql = \"\"\"INSERT INTO germplasm (Locus, germplasm, phenotype, pubmed) values ('{}', '{}', '{}', '{}')\"\"\".format(row[\"Locus\"], row[\"germplasm\"], row[\"phenotype\"], row[\"pubmed\"])\n",
    "                cursor.execute(sql)\n",
    "        connection.commit() \n",
    "finally: \n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OrderedDict([('Locus', 'AT1G01040'), ('Gene', 'DCL1'), ('ProteinLength', '332')])\n",
      "OrderedDict([('Locus', 'AT1G01060'), ('Gene', 'LHY'), ('ProteinLength', '290')])\n",
      "OrderedDict([('Locus', 'AT1G01140'), ('Gene', 'CIPK9'), ('ProteinLength', '223')])\n",
      "OrderedDict([('Locus', 'AT1G01220'), ('Gene', 'FKGP'), ('ProteinLength', '190')])\n",
      "OrderedDict([('Locus', 'AT2G03720'), ('Gene', 'MRH6'), ('ProteinLength', '189')])\n",
      "OrderedDict([('Locus', 'AT2G03800'), ('Gene', 'GEK1'), ('ProteinLength', '196')])\n",
      "OrderedDict([('Locus', 'AT2G04240'), ('Gene', 'XERICO'), ('ProteinLength', '256')])\n",
      "OrderedDict([('Locus', 'AT2G05210'), ('Gene', 'POT1A'), ('ProteinLength', '221')])\n",
      "OrderedDict([('Locus', 'AT3G02130'), ('Gene', 'RPK2'), ('ProteinLength', '284')])\n",
      "OrderedDict([('Locus', 'AT3G02140'), ('Gene', 'TMAC2'), ('ProteinLength', '300')])\n",
      "OrderedDict([('Locus', 'AT3G02230'), ('Gene', 'RGP1'), ('ProteinLength', '301')])\n",
      "OrderedDict([('Locus', 'AT3G02260'), ('Gene', 'BIG'), ('ProteinLength', '279')])\n",
      "OrderedDict([('Locus', 'AT3G02310'), ('Gene', 'SEP2'), ('ProteinLength', '175')])\n",
      "OrderedDict([('Locus', 'AT3G02680'), ('Gene', 'NBS1'), ('ProteinLength', '190')])\n",
      "OrderedDict([('Locus', 'AT3G02850'), ('Gene', 'SKOR'), ('ProteinLength', '234')])\n",
      "OrderedDict([('Locus', 'AT3G02870'), ('Gene', 'VTC4'), ('ProteinLength', '311')])\n",
      "OrderedDict([('Locus', 'AT3G03260'), ('Gene', 'HDG8'), ('ProteinLength', '194')])\n",
      "OrderedDict([('Locus', 'AT4G14790'), ('Gene', 'SUV3'), ('ProteinLength', '312')])\n",
      "OrderedDict([('Locus', 'AT4G15210'), ('Gene', 'BAM5'), ('ProteinLength', '313')])\n",
      "OrderedDict([('Locus', 'AT4G15560'), ('Gene', 'DXS'), ('ProteinLength', '219')])\n",
      "OrderedDict([('Locus', 'AT4G15570'), ('Gene', 'MAA3'), ('ProteinLength', '294')])\n",
      "OrderedDict([('Locus', 'AT4G15802'), ('Gene', 'HSBP'), ('ProteinLength', '254')])\n",
      "OrderedDict([('Locus', 'AT4G15880'), ('Gene', 'ESD4'), ('ProteinLength', '265')])\n",
      "OrderedDict([('Locus', 'AT4G16420'), ('Gene', 'ADA2B'), ('ProteinLength', '279')])\n",
      "OrderedDict([('Locus', 'AT4G16480'), ('Gene', 'INT4'), ('ProteinLength', '284')])\n",
      "OrderedDict([('Locus', 'AT5G10480'), ('Gene', 'PAS2'), ('ProteinLength', '301')])\n",
      "OrderedDict([('Locus', 'AT5G10510'), ('Gene', 'AIL6'), ('ProteinLength', '310')])\n",
      "OrderedDict([('Locus', 'AT5G11110'), ('Gene', 'SPS2'), ('ProteinLength', '232')])\n",
      "OrderedDict([('Locus', 'AT5G11260'), ('Gene', 'HY5'), ('ProteinLength', '221')])\n",
      "OrderedDict([('Locus', 'AT5G11510'), ('Gene', 'MYB3R-4'), ('ProteinLength', '336')])\n",
      "OrderedDict([('Locus', 'AT5G12200'), ('Gene', 'PYD2'), ('ProteinLength', '310')])\n",
      "OrderedDict([('Locus', 'AT5G13290'), ('Gene', 'CRN'), ('ProteinLength', '189')])\n"
     ]
    }
   ],
   "source": [
    "#insert data in the table of data base\n",
    "import pymysql.cursors\n",
    "import csv\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                            user='root',\n",
    "                            password='root',\n",
    "                            db='exam2',\n",
    "                            charset='utf8mb4', \n",
    "                            cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    " \n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open('LocusGene.tsv') as csvfile:\n",
    "            locus = csv.DictReader(csvfile, delimiter = '\\t')\n",
    "            for row in locus:\n",
    "                print(row)\n",
    "                sql = \"\"\"INSERT INTO locus (Locus, Gene, ProteinLength) values ('{}', '{}', '{}')\"\"\".format(row[\"Locus\"], row[\"Gene\"], row[\"ProteinLength\"])\n",
    "                cursor.execute(sql)\n",
    "        connection.commit() \n",
    "finally: \n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "32 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>id</th>\n",
       "        <th>Locus</th>\n",
       "        <th>Gene</th>\n",
       "        <th>ProteinLength</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>1</td>\n",
       "        <td>AT1G01040</td>\n",
       "        <td>DCL1</td>\n",
       "        <td>332.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>2</td>\n",
       "        <td>AT1G01060</td>\n",
       "        <td>LHY</td>\n",
       "        <td>290.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>3</td>\n",
       "        <td>AT1G01140</td>\n",
       "        <td>CIPK9</td>\n",
       "        <td>223.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>4</td>\n",
       "        <td>AT1G01220</td>\n",
       "        <td>FKGP</td>\n",
       "        <td>190.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>5</td>\n",
       "        <td>AT2G03720</td>\n",
       "        <td>MRH6</td>\n",
       "        <td>189.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>6</td>\n",
       "        <td>AT2G03800</td>\n",
       "        <td>GEK1</td>\n",
       "        <td>196.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>7</td>\n",
       "        <td>AT2G04240</td>\n",
       "        <td>XERICO</td>\n",
       "        <td>256.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>8</td>\n",
       "        <td>AT2G05210</td>\n",
       "        <td>POT1A</td>\n",
       "        <td>221.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>9</td>\n",
       "        <td>AT3G02130</td>\n",
       "        <td>RPK2</td>\n",
       "        <td>284.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>10</td>\n",
       "        <td>AT3G02140</td>\n",
       "        <td>TMAC2</td>\n",
       "        <td>300.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>11</td>\n",
       "        <td>AT3G02230</td>\n",
       "        <td>RGP1</td>\n",
       "        <td>301.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>12</td>\n",
       "        <td>AT3G02260</td>\n",
       "        <td>BIG</td>\n",
       "        <td>279.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>13</td>\n",
       "        <td>AT3G02310</td>\n",
       "        <td>SEP2</td>\n",
       "        <td>175.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>14</td>\n",
       "        <td>AT3G02680</td>\n",
       "        <td>NBS1</td>\n",
       "        <td>190.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>15</td>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>SKOR</td>\n",
       "        <td>234.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>16</td>\n",
       "        <td>AT3G02870</td>\n",
       "        <td>VTC4</td>\n",
       "        <td>311.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>17</td>\n",
       "        <td>AT3G03260</td>\n",
       "        <td>HDG8</td>\n",
       "        <td>194.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>18</td>\n",
       "        <td>AT4G14790</td>\n",
       "        <td>SUV3</td>\n",
       "        <td>312.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>19</td>\n",
       "        <td>AT4G15210</td>\n",
       "        <td>BAM5</td>\n",
       "        <td>313.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>20</td>\n",
       "        <td>AT4G15560</td>\n",
       "        <td>DXS</td>\n",
       "        <td>219.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>21</td>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>MAA3</td>\n",
       "        <td>294.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>22</td>\n",
       "        <td>AT4G15802</td>\n",
       "        <td>HSBP</td>\n",
       "        <td>254.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>23</td>\n",
       "        <td>AT4G15880</td>\n",
       "        <td>ESD4</td>\n",
       "        <td>265.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>24</td>\n",
       "        <td>AT4G16420</td>\n",
       "        <td>ADA2B</td>\n",
       "        <td>279.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>25</td>\n",
       "        <td>AT4G16480</td>\n",
       "        <td>INT4</td>\n",
       "        <td>284.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>26</td>\n",
       "        <td>AT5G10480</td>\n",
       "        <td>PAS2</td>\n",
       "        <td>301.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>27</td>\n",
       "        <td>AT5G10510</td>\n",
       "        <td>AIL6</td>\n",
       "        <td>310.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>28</td>\n",
       "        <td>AT5G11110</td>\n",
       "        <td>SPS2</td>\n",
       "        <td>232.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>29</td>\n",
       "        <td>AT5G11260</td>\n",
       "        <td>HY5</td>\n",
       "        <td>221.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>30</td>\n",
       "        <td>AT5G11510</td>\n",
       "        <td>MYB3R-4</td>\n",
       "        <td>336.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>31</td>\n",
       "        <td>AT5G12200</td>\n",
       "        <td>PYD2</td>\n",
       "        <td>310.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>32</td>\n",
       "        <td>AT5G13290</td>\n",
       "        <td>CRN</td>\n",
       "        <td>189.0</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[(1, 'AT1G01040', 'DCL1', 332.0),\n",
       " (2, 'AT1G01060', 'LHY', 290.0),\n",
       " (3, 'AT1G01140', 'CIPK9', 223.0),\n",
       " (4, 'AT1G01220', 'FKGP', 190.0),\n",
       " (5, 'AT2G03720', 'MRH6', 189.0),\n",
       " (6, 'AT2G03800', 'GEK1', 196.0),\n",
       " (7, 'AT2G04240', 'XERICO', 256.0),\n",
       " (8, 'AT2G05210', 'POT1A', 221.0),\n",
       " (9, 'AT3G02130', 'RPK2', 284.0),\n",
       " (10, 'AT3G02140', 'TMAC2', 300.0),\n",
       " (11, 'AT3G02230', 'RGP1', 301.0),\n",
       " (12, 'AT3G02260', 'BIG', 279.0),\n",
       " (13, 'AT3G02310', 'SEP2', 175.0),\n",
       " (14, 'AT3G02680', 'NBS1', 190.0),\n",
       " (15, 'AT3G02850', 'SKOR', 234.0),\n",
       " (16, 'AT3G02870', 'VTC4', 311.0),\n",
       " (17, 'AT3G03260', 'HDG8', 194.0),\n",
       " (18, 'AT4G14790', 'SUV3', 312.0),\n",
       " (19, 'AT4G15210', 'BAM5', 313.0),\n",
       " (20, 'AT4G15560', 'DXS', 219.0),\n",
       " (21, 'AT4G15570', 'MAA3', 294.0),\n",
       " (22, 'AT4G15802', 'HSBP', 254.0),\n",
       " (23, 'AT4G15880', 'ESD4', 265.0),\n",
       " (24, 'AT4G16420', 'ADA2B', 279.0),\n",
       " (25, 'AT4G16480', 'INT4', 284.0),\n",
       " (26, 'AT5G10480', 'PAS2', 301.0),\n",
       " (27, 'AT5G10510', 'AIL6', 310.0),\n",
       " (28, 'AT5G11110', 'SPS2', 232.0),\n",
       " (29, 'AT5G11260', 'HY5', 221.0),\n",
       " (30, 'AT5G11510', 'MYB3R-4', 336.0),\n",
       " (31, 'AT5G12200', 'PYD2', 310.0),\n",
       " (32, 'AT5G13290', 'CRN', 189.0)]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#probe the result in the table\n",
    "%sql SELECT * FROM locus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "32 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>id</th>\n",
       "        <th>Locus</th>\n",
       "        <th>germplasm</th>\n",
       "        <th>phenotype</th>\n",
       "        <th>pubmed</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>1</td>\n",
       "        <td>AT1G01040</td>\n",
       "        <td>CS3828</td>\n",
       "        <td>Increased abundance of miRNA precursors.</td>\n",
       "        <td>17369351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>2</td>\n",
       "        <td>AT1G01060</td>\n",
       "        <td>lhy-101</td>\n",
       "        <td>The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.</td>\n",
       "        <td>16891401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>3</td>\n",
       "        <td>AT1G01140</td>\n",
       "        <td>SALK_058629</td>\n",
       "        <td>hypersensitive to low potassium media</td>\n",
       "        <td>17486125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>4</td>\n",
       "        <td>AT1G01220</td>\n",
       "        <td>SALK_012400C</td>\n",
       "        <td>fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.</td>\n",
       "        <td>18199744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>5</td>\n",
       "        <td>AT2G03720</td>\n",
       "        <td>SALK_042433</td>\n",
       "        <td>Multiple straight hairs</td>\n",
       "        <td>16367956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>6</td>\n",
       "        <td>AT2G03800</td>\n",
       "        <td>gek1-1</td>\n",
       "        <td>Ethanol hypersensitivity.</td>\n",
       "        <td>15215505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>7</td>\n",
       "        <td>AT2G04240</td>\n",
       "        <td>xerico</td>\n",
       "        <td>Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.</td>\n",
       "        <td>17933900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>8</td>\n",
       "        <td>AT2G05210</td>\n",
       "        <td>pot1-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>17627276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>9</td>\n",
       "        <td>AT3G02130</td>\n",
       "        <td>rpk2-2</td>\n",
       "        <td>The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.</td>\n",
       "        <td>17419837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>10</td>\n",
       "        <td>AT3G02140</td>\n",
       "        <td>afp4-1</td>\n",
       "        <td>Decreased germination on high concentrations of glucose and sorbitol.</td>\n",
       "        <td>18484180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>11</td>\n",
       "        <td>AT3G02230</td>\n",
       "        <td>rgp1-1</td>\n",
       "        <td>rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.</td>\n",
       "        <td>21478444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>12</td>\n",
       "        <td>AT3G02260</td>\n",
       "        <td>tir3-1 RGLG1:rglg1 rglg2</td>\n",
       "        <td>The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.</td>\n",
       "        <td>17586653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>13</td>\n",
       "        <td>AT3G02310</td>\n",
       "        <td>sep2-1</td>\n",
       "        <td>Non-described subtle phenotype.</td>\n",
       "        <td>10821278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>14</td>\n",
       "        <td>AT3G02680</td>\n",
       "        <td>atnbs1-1</td>\n",
       "        <td>Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.</td>\n",
       "        <td>17672843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>15</td>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>CS3816</td>\n",
       "        <td>The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.</td>\n",
       "        <td>17568770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>16</td>\n",
       "        <td>AT3G02870</td>\n",
       "        <td>vtc4-1</td>\n",
       "        <td>ascorbate deficient</td>\n",
       "        <td>16595667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>17</td>\n",
       "        <td>AT3G03260</td>\n",
       "        <td>hdg8-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>16778018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>18</td>\n",
       "        <td>AT4G14790</td>\n",
       "        <td>pdd17</td>\n",
       "        <td>Defective pollen development.</td>\n",
       "        <td>19237690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>19</td>\n",
       "        <td>AT4G15210</td>\n",
       "        <td>bmy1-2</td>\n",
       "        <td>Plants cold-shocked for 6h have an increased starch content compared to wildtype.</td>\n",
       "        <td>16297066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>20</td>\n",
       "        <td>AT4G15560</td>\n",
       "        <td>cla1-1</td>\n",
       "        <td>Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.</td>\n",
       "        <td>10982425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>21</td>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>maa3</td>\n",
       "        <td>Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.</td>\n",
       "        <td>18772186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>22</td>\n",
       "        <td>AT4G15802</td>\n",
       "        <td>Athspb-2</td>\n",
       "        <td>Early flowering, reduced fertility, aborted seeds.</td>\n",
       "        <td>20388662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>23</td>\n",
       "        <td>AT4G15880</td>\n",
       "        <td>esd4-2</td>\n",
       "        <td>Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.</td>\n",
       "        <td>17513499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>24</td>\n",
       "        <td>AT4G16420</td>\n",
       "        <td>prz1-1</td>\n",
       "        <td>Altered response to auxin and cytokinin</td>\n",
       "        <td>12747832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>25</td>\n",
       "        <td>AT4G16480</td>\n",
       "        <td>atint4-2</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>16603666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>26</td>\n",
       "        <td>AT5G10480</td>\n",
       "        <td>pas2-3</td>\n",
       "        <td>Segregates 25% embryo lethal.</td>\n",
       "        <td>18799749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>27</td>\n",
       "        <td>AT5G10510</td>\n",
       "        <td>plt3-1</td>\n",
       "        <td>Short roots and shortened root meristem.</td>\n",
       "        <td>17960244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>28</td>\n",
       "        <td>AT5G11110</td>\n",
       "        <td>kns2</td>\n",
       "        <td>Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.</td>\n",
       "        <td>18779216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>29</td>\n",
       "        <td>AT5G11260</td>\n",
       "        <td>hy5-101</td>\n",
       "        <td>Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.</td>\n",
       "        <td>16891401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>30</td>\n",
       "        <td>AT5G11510</td>\n",
       "        <td>myb3r4-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>17287251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>31</td>\n",
       "        <td>AT5G12200</td>\n",
       "        <td>pyd2-2</td>\n",
       "        <td>The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.</td>\n",
       "        <td>19413687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>32</td>\n",
       "        <td>AT5G13290</td>\n",
       "        <td>crn-1</td>\n",
       "        <td>Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.</td>\n",
       "        <td>12345678</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[(1, 'AT1G01040', 'CS3828', 'Increased abundance of miRNA precursors.', '17369351'),\n",
       " (2, 'AT1G01060', 'lhy-101', 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.', '16891401'),\n",
       " (3, 'AT1G01140', 'SALK_058629', 'hypersensitive to low potassium media', '17486125'),\n",
       " (4, 'AT1G01220', 'SALK_012400C', 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ s ... (41 characters truncated) ... henotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.', '18199744'),\n",
       " (5, 'AT2G03720', 'SALK_042433', 'Multiple straight hairs', '16367956'),\n",
       " (6, 'AT2G03800', 'gek1-1', 'Ethanol hypersensitivity.', '15215505'),\n",
       " (7, 'AT2G04240', 'xerico', 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.', '17933900'),\n",
       " (8, 'AT2G05210', 'pot1-1', 'No visible phenotype.', '17627276'),\n",
       " (9, 'AT3G02130', 'rpk2-2', 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.', '17419837'),\n",
       " (10, 'AT3G02140', 'afp4-1', 'Decreased germination on high concentrations of glucose and sorbitol.', '18484180'),\n",
       " (11, 'AT3G02230', 'rgp1-1', 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.', '21478444'),\n",
       " (12, 'AT3G02260', 'tir3-1 RGLG1:rglg1 rglg2', 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.', '17586653'),\n",
       " (13, 'AT3G02310', 'sep2-1', 'Non-described subtle phenotype.', '10821278'),\n",
       " (14, 'AT3G02680', 'atnbs1-1', 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.', '17672843'),\n",
       " (15, 'AT3G02850', 'CS3816', 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', '17568770'),\n",
       " (16, 'AT3G02870', 'vtc4-1', 'ascorbate deficient', '16595667'),\n",
       " (17, 'AT3G03260', 'hdg8-1', 'No visible phenotype.', '16778018'),\n",
       " (18, 'AT4G14790', 'pdd17', 'Defective pollen development.', '19237690'),\n",
       " (19, 'AT4G15210', 'bmy1-2', 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.', '16297066'),\n",
       " (20, 'AT4G15560', 'cla1-1', 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.', '10982425'),\n",
       " (21, 'AT4G15570', 'maa3', 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', '18772186'),\n",
       " (22, 'AT4G15802', 'Athspb-2', 'Early flowering, reduced fertility, aborted seeds.', '20388662'),\n",
       " (23, 'AT4G15880', 'esd4-2', 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.', '17513499'),\n",
       " (24, 'AT4G16420', 'prz1-1', 'Altered response to auxin and cytokinin', '12747832'),\n",
       " (25, 'AT4G16480', 'atint4-2', 'No visible phenotype.', '16603666'),\n",
       " (26, 'AT5G10480', 'pas2-3', 'Segregates 25% embryo lethal.', '18799749'),\n",
       " (27, 'AT5G10510', 'plt3-1', 'Short roots and shortened root meristem.', '17960244'),\n",
       " (28, 'AT5G11110', 'kns2', 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.', '18779216'),\n",
       " (29, 'AT5G11260', 'hy5-101', 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.', '16891401'),\n",
       " (30, 'AT5G11510', 'myb3r4-1', 'No visible phenotype.', '17287251'),\n",
       " (31, 'AT5G12200', 'pyd2-2', 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.', '19413687'),\n",
       " (32, 'AT5G13290', 'crn-1', 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.', '12345678')]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#probe the reult in the table\n",
    "%sql SELECT * FROM germplasm;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "32 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>id</th>\n",
       "        <th>Locus</th>\n",
       "        <th>germplasm</th>\n",
       "        <th>phenotype</th>\n",
       "        <th>pubmed</th>\n",
       "        <th>id_1</th>\n",
       "        <th>Locus_1</th>\n",
       "        <th>Gene</th>\n",
       "        <th>ProteinLength</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>1</td>\n",
       "        <td>AT1G01040</td>\n",
       "        <td>CS3828</td>\n",
       "        <td>Increased abundance of miRNA precursors.</td>\n",
       "        <td>17369351</td>\n",
       "        <td>1</td>\n",
       "        <td>AT1G01040</td>\n",
       "        <td>DCL1</td>\n",
       "        <td>332.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>2</td>\n",
       "        <td>AT1G01060</td>\n",
       "        <td>lhy-101</td>\n",
       "        <td>The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.</td>\n",
       "        <td>16891401</td>\n",
       "        <td>2</td>\n",
       "        <td>AT1G01060</td>\n",
       "        <td>LHY</td>\n",
       "        <td>290.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>3</td>\n",
       "        <td>AT1G01140</td>\n",
       "        <td>SALK_058629</td>\n",
       "        <td>hypersensitive to low potassium media</td>\n",
       "        <td>17486125</td>\n",
       "        <td>3</td>\n",
       "        <td>AT1G01140</td>\n",
       "        <td>CIPK9</td>\n",
       "        <td>223.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>4</td>\n",
       "        <td>AT1G01220</td>\n",
       "        <td>SALK_012400C</td>\n",
       "        <td>fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.</td>\n",
       "        <td>18199744</td>\n",
       "        <td>4</td>\n",
       "        <td>AT1G01220</td>\n",
       "        <td>FKGP</td>\n",
       "        <td>190.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>5</td>\n",
       "        <td>AT2G03720</td>\n",
       "        <td>SALK_042433</td>\n",
       "        <td>Multiple straight hairs</td>\n",
       "        <td>16367956</td>\n",
       "        <td>5</td>\n",
       "        <td>AT2G03720</td>\n",
       "        <td>MRH6</td>\n",
       "        <td>189.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>6</td>\n",
       "        <td>AT2G03800</td>\n",
       "        <td>gek1-1</td>\n",
       "        <td>Ethanol hypersensitivity.</td>\n",
       "        <td>15215505</td>\n",
       "        <td>6</td>\n",
       "        <td>AT2G03800</td>\n",
       "        <td>GEK1</td>\n",
       "        <td>196.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>7</td>\n",
       "        <td>AT2G04240</td>\n",
       "        <td>xerico</td>\n",
       "        <td>Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.</td>\n",
       "        <td>17933900</td>\n",
       "        <td>7</td>\n",
       "        <td>AT2G04240</td>\n",
       "        <td>XERICO</td>\n",
       "        <td>256.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>8</td>\n",
       "        <td>AT2G05210</td>\n",
       "        <td>pot1-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>17627276</td>\n",
       "        <td>8</td>\n",
       "        <td>AT2G05210</td>\n",
       "        <td>POT1A</td>\n",
       "        <td>221.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>9</td>\n",
       "        <td>AT3G02130</td>\n",
       "        <td>rpk2-2</td>\n",
       "        <td>The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.</td>\n",
       "        <td>17419837</td>\n",
       "        <td>9</td>\n",
       "        <td>AT3G02130</td>\n",
       "        <td>RPK2</td>\n",
       "        <td>284.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>10</td>\n",
       "        <td>AT3G02140</td>\n",
       "        <td>afp4-1</td>\n",
       "        <td>Decreased germination on high concentrations of glucose and sorbitol.</td>\n",
       "        <td>18484180</td>\n",
       "        <td>10</td>\n",
       "        <td>AT3G02140</td>\n",
       "        <td>TMAC2</td>\n",
       "        <td>300.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>11</td>\n",
       "        <td>AT3G02230</td>\n",
       "        <td>rgp1-1</td>\n",
       "        <td>rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.</td>\n",
       "        <td>21478444</td>\n",
       "        <td>11</td>\n",
       "        <td>AT3G02230</td>\n",
       "        <td>RGP1</td>\n",
       "        <td>301.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>12</td>\n",
       "        <td>AT3G02260</td>\n",
       "        <td>tir3-1 RGLG1:rglg1 rglg2</td>\n",
       "        <td>The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.</td>\n",
       "        <td>17586653</td>\n",
       "        <td>12</td>\n",
       "        <td>AT3G02260</td>\n",
       "        <td>BIG</td>\n",
       "        <td>279.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>13</td>\n",
       "        <td>AT3G02310</td>\n",
       "        <td>sep2-1</td>\n",
       "        <td>Non-described subtle phenotype.</td>\n",
       "        <td>10821278</td>\n",
       "        <td>13</td>\n",
       "        <td>AT3G02310</td>\n",
       "        <td>SEP2</td>\n",
       "        <td>175.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>14</td>\n",
       "        <td>AT3G02680</td>\n",
       "        <td>atnbs1-1</td>\n",
       "        <td>Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.</td>\n",
       "        <td>17672843</td>\n",
       "        <td>14</td>\n",
       "        <td>AT3G02680</td>\n",
       "        <td>NBS1</td>\n",
       "        <td>190.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>15</td>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>CS3816</td>\n",
       "        <td>The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.</td>\n",
       "        <td>17568770</td>\n",
       "        <td>15</td>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>SKOR</td>\n",
       "        <td>234.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>16</td>\n",
       "        <td>AT3G02870</td>\n",
       "        <td>vtc4-1</td>\n",
       "        <td>ascorbate deficient</td>\n",
       "        <td>16595667</td>\n",
       "        <td>16</td>\n",
       "        <td>AT3G02870</td>\n",
       "        <td>VTC4</td>\n",
       "        <td>311.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>17</td>\n",
       "        <td>AT3G03260</td>\n",
       "        <td>hdg8-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>16778018</td>\n",
       "        <td>17</td>\n",
       "        <td>AT3G03260</td>\n",
       "        <td>HDG8</td>\n",
       "        <td>194.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>18</td>\n",
       "        <td>AT4G14790</td>\n",
       "        <td>pdd17</td>\n",
       "        <td>Defective pollen development.</td>\n",
       "        <td>19237690</td>\n",
       "        <td>18</td>\n",
       "        <td>AT4G14790</td>\n",
       "        <td>SUV3</td>\n",
       "        <td>312.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>19</td>\n",
       "        <td>AT4G15210</td>\n",
       "        <td>bmy1-2</td>\n",
       "        <td>Plants cold-shocked for 6h have an increased starch content compared to wildtype.</td>\n",
       "        <td>16297066</td>\n",
       "        <td>19</td>\n",
       "        <td>AT4G15210</td>\n",
       "        <td>BAM5</td>\n",
       "        <td>313.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>20</td>\n",
       "        <td>AT4G15560</td>\n",
       "        <td>cla1-1</td>\n",
       "        <td>Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.</td>\n",
       "        <td>10982425</td>\n",
       "        <td>20</td>\n",
       "        <td>AT4G15560</td>\n",
       "        <td>DXS</td>\n",
       "        <td>219.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>21</td>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>maa3</td>\n",
       "        <td>Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.</td>\n",
       "        <td>18772186</td>\n",
       "        <td>21</td>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>MAA3</td>\n",
       "        <td>294.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>22</td>\n",
       "        <td>AT4G15802</td>\n",
       "        <td>Athspb-2</td>\n",
       "        <td>Early flowering, reduced fertility, aborted seeds.</td>\n",
       "        <td>20388662</td>\n",
       "        <td>22</td>\n",
       "        <td>AT4G15802</td>\n",
       "        <td>HSBP</td>\n",
       "        <td>254.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>23</td>\n",
       "        <td>AT4G15880</td>\n",
       "        <td>esd4-2</td>\n",
       "        <td>Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.</td>\n",
       "        <td>17513499</td>\n",
       "        <td>23</td>\n",
       "        <td>AT4G15880</td>\n",
       "        <td>ESD4</td>\n",
       "        <td>265.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>24</td>\n",
       "        <td>AT4G16420</td>\n",
       "        <td>prz1-1</td>\n",
       "        <td>Altered response to auxin and cytokinin</td>\n",
       "        <td>12747832</td>\n",
       "        <td>24</td>\n",
       "        <td>AT4G16420</td>\n",
       "        <td>ADA2B</td>\n",
       "        <td>279.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>25</td>\n",
       "        <td>AT4G16480</td>\n",
       "        <td>atint4-2</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>16603666</td>\n",
       "        <td>25</td>\n",
       "        <td>AT4G16480</td>\n",
       "        <td>INT4</td>\n",
       "        <td>284.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>26</td>\n",
       "        <td>AT5G10480</td>\n",
       "        <td>pas2-3</td>\n",
       "        <td>Segregates 25% embryo lethal.</td>\n",
       "        <td>18799749</td>\n",
       "        <td>26</td>\n",
       "        <td>AT5G10480</td>\n",
       "        <td>PAS2</td>\n",
       "        <td>301.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>27</td>\n",
       "        <td>AT5G10510</td>\n",
       "        <td>plt3-1</td>\n",
       "        <td>Short roots and shortened root meristem.</td>\n",
       "        <td>17960244</td>\n",
       "        <td>27</td>\n",
       "        <td>AT5G10510</td>\n",
       "        <td>AIL6</td>\n",
       "        <td>310.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>28</td>\n",
       "        <td>AT5G11110</td>\n",
       "        <td>kns2</td>\n",
       "        <td>Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.</td>\n",
       "        <td>18779216</td>\n",
       "        <td>28</td>\n",
       "        <td>AT5G11110</td>\n",
       "        <td>SPS2</td>\n",
       "        <td>232.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>29</td>\n",
       "        <td>AT5G11260</td>\n",
       "        <td>hy5-101</td>\n",
       "        <td>Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.</td>\n",
       "        <td>16891401</td>\n",
       "        <td>29</td>\n",
       "        <td>AT5G11260</td>\n",
       "        <td>HY5</td>\n",
       "        <td>221.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>30</td>\n",
       "        <td>AT5G11510</td>\n",
       "        <td>myb3r4-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>17287251</td>\n",
       "        <td>30</td>\n",
       "        <td>AT5G11510</td>\n",
       "        <td>MYB3R-4</td>\n",
       "        <td>336.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>31</td>\n",
       "        <td>AT5G12200</td>\n",
       "        <td>pyd2-2</td>\n",
       "        <td>The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.</td>\n",
       "        <td>19413687</td>\n",
       "        <td>31</td>\n",
       "        <td>AT5G12200</td>\n",
       "        <td>PYD2</td>\n",
       "        <td>310.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>32</td>\n",
       "        <td>AT5G13290</td>\n",
       "        <td>crn-1</td>\n",
       "        <td>Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.</td>\n",
       "        <td>12345678</td>\n",
       "        <td>32</td>\n",
       "        <td>AT5G13290</td>\n",
       "        <td>CRN</td>\n",
       "        <td>189.0</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[(1, 'AT1G01040', 'CS3828', 'Increased abundance of miRNA precursors.', '17369351', 1, 'AT1G01040', 'DCL1', 332.0),\n",
       " (2, 'AT1G01060', 'lhy-101', 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.', '16891401', 2, 'AT1G01060', 'LHY', 290.0),\n",
       " (3, 'AT1G01140', 'SALK_058629', 'hypersensitive to low potassium media', '17486125', 3, 'AT1G01140', 'CIPK9', 223.0),\n",
       " (4, 'AT1G01220', 'SALK_012400C', 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ s ... (41 characters truncated) ... henotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.', '18199744', 4, 'AT1G01220', 'FKGP', 190.0),\n",
       " (5, 'AT2G03720', 'SALK_042433', 'Multiple straight hairs', '16367956', 5, 'AT2G03720', 'MRH6', 189.0),\n",
       " (6, 'AT2G03800', 'gek1-1', 'Ethanol hypersensitivity.', '15215505', 6, 'AT2G03800', 'GEK1', 196.0),\n",
       " (7, 'AT2G04240', 'xerico', 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.', '17933900', 7, 'AT2G04240', 'XERICO', 256.0),\n",
       " (8, 'AT2G05210', 'pot1-1', 'No visible phenotype.', '17627276', 8, 'AT2G05210', 'POT1A', 221.0),\n",
       " (9, 'AT3G02130', 'rpk2-2', 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.', '17419837', 9, 'AT3G02130', 'RPK2', 284.0),\n",
       " (10, 'AT3G02140', 'afp4-1', 'Decreased germination on high concentrations of glucose and sorbitol.', '18484180', 10, 'AT3G02140', 'TMAC2', 300.0),\n",
       " (11, 'AT3G02230', 'rgp1-1', 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.', '21478444', 11, 'AT3G02230', 'RGP1', 301.0),\n",
       " (12, 'AT3G02260', 'tir3-1 RGLG1:rglg1 rglg2', 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.', '17586653', 12, 'AT3G02260', 'BIG', 279.0),\n",
       " (13, 'AT3G02310', 'sep2-1', 'Non-described subtle phenotype.', '10821278', 13, 'AT3G02310', 'SEP2', 175.0),\n",
       " (14, 'AT3G02680', 'atnbs1-1', 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.', '17672843', 14, 'AT3G02680', 'NBS1', 190.0),\n",
       " (15, 'AT3G02850', 'CS3816', 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', '17568770', 15, 'AT3G02850', 'SKOR', 234.0),\n",
       " (16, 'AT3G02870', 'vtc4-1', 'ascorbate deficient', '16595667', 16, 'AT3G02870', 'VTC4', 311.0),\n",
       " (17, 'AT3G03260', 'hdg8-1', 'No visible phenotype.', '16778018', 17, 'AT3G03260', 'HDG8', 194.0),\n",
       " (18, 'AT4G14790', 'pdd17', 'Defective pollen development.', '19237690', 18, 'AT4G14790', 'SUV3', 312.0),\n",
       " (19, 'AT4G15210', 'bmy1-2', 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.', '16297066', 19, 'AT4G15210', 'BAM5', 313.0),\n",
       " (20, 'AT4G15560', 'cla1-1', 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.', '10982425', 20, 'AT4G15560', 'DXS', 219.0),\n",
       " (21, 'AT4G15570', 'maa3', 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', '18772186', 21, 'AT4G15570', 'MAA3', 294.0),\n",
       " (22, 'AT4G15802', 'Athspb-2', 'Early flowering, reduced fertility, aborted seeds.', '20388662', 22, 'AT4G15802', 'HSBP', 254.0),\n",
       " (23, 'AT4G15880', 'esd4-2', 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.', '17513499', 23, 'AT4G15880', 'ESD4', 265.0),\n",
       " (24, 'AT4G16420', 'prz1-1', 'Altered response to auxin and cytokinin', '12747832', 24, 'AT4G16420', 'ADA2B', 279.0),\n",
       " (25, 'AT4G16480', 'atint4-2', 'No visible phenotype.', '16603666', 25, 'AT4G16480', 'INT4', 284.0),\n",
       " (26, 'AT5G10480', 'pas2-3', 'Segregates 25% embryo lethal.', '18799749', 26, 'AT5G10480', 'PAS2', 301.0),\n",
       " (27, 'AT5G10510', 'plt3-1', 'Short roots and shortened root meristem.', '17960244', 27, 'AT5G10510', 'AIL6', 310.0),\n",
       " (28, 'AT5G11110', 'kns2', 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.', '18779216', 28, 'AT5G11110', 'SPS2', 232.0),\n",
       " (29, 'AT5G11260', 'hy5-101', 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.', '16891401', 29, 'AT5G11260', 'HY5', 221.0),\n",
       " (30, 'AT5G11510', 'myb3r4-1', 'No visible phenotype.', '17287251', 30, 'AT5G11510', 'MYB3R-4', 336.0),\n",
       " (31, 'AT5G12200', 'pyd2-2', 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.', '19413687', 31, 'AT5G12200', 'PYD2', 310.0),\n",
       " (32, 'AT5G13290', 'crn-1', 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.', '12345678', 32, 'AT5G13290', 'CRN', 189.0)]"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# selection of join of two tables\n",
    "exe1 = %sql SELECT * FROM germplasm RIGHT JOIN locus ON germplasm.Locus = locus.Locus;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "2 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>id</th>\n",
       "        <th>Locus</th>\n",
       "        <th>germplasm</th>\n",
       "        <th>phenotype</th>\n",
       "        <th>pubmed</th>\n",
       "        <th>id_1</th>\n",
       "        <th>Locus_1</th>\n",
       "        <th>Gene</th>\n",
       "        <th>ProteinLength</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>15</td>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>CS3816</td>\n",
       "        <td>The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.</td>\n",
       "        <td>17568770</td>\n",
       "        <td>15</td>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>SKOR</td>\n",
       "        <td>234.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>21</td>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>maa3</td>\n",
       "        <td>Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.</td>\n",
       "        <td>18772186</td>\n",
       "        <td>21</td>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>MAA3</td>\n",
       "        <td>294.0</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[(15, 'AT3G02850', 'CS3816', 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', '17568770', 15, 'AT3G02850', 'SKOR', 234.0),\n",
       " (21, 'AT4G15570', 'maa3', 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', '18772186', 21, 'AT4G15570', 'MAA3', 294.0)]"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#SElection of gene skor and maa3\n",
    "exe2 = %sql SELECT * FROM germplasm RIGHT JOIN locus ON germplasm.Locus = locus.Locus WHERE locus.Gene = \"SKOR\" OR locus.Gene = \"MAA3\";"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "can only concatenate str (not \"int\") to str",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-142-2d024c7a7903>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0mx\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"AT\"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m\"%\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mexe\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun_line_magic\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'sql'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'SELECT COUNT(*) FROM germplasm RIGHT JOIN locus ON germplasm.Locus = locus.Locus WHERE germplasm.Locus LIKE :x ;  '\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m     \u001b[0mexe3\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"Chromosome\"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mn\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mexe\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      7\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Chromosome\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexe\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexe3\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: can only concatenate str (not \"int\") to str"
     ]
    }
   ],
   "source": [
    "#Selection by chromosome\n",
    "exe3 = {}\n",
    "for n in range(1,6):\n",
    "    x = \"AT\" + str(n) + \"%\"\n",
    "    exe = %sql SELECT COUNT(*) FROM germplasm RIGHT JOIN locus ON germplasm.Locus = locus.Locus WHERE germplasm.Locus LIKE :x ;  \n",
    "    exe3[n] = exe\n",
    "    print(\"Chromosome\", n, exe)\n",
    "print(exe3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      "Chromosome 1 +--------------------------+\n",
      "| AVG(locus.ProteinLength) |\n",
      "+--------------------------+\n",
      "|          258.75          |\n",
      "+--------------------------+\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      "Chromosome 2 +--------------------------+\n",
      "| AVG(locus.ProteinLength) |\n",
      "+--------------------------+\n",
      "|          215.5           |\n",
      "+--------------------------+\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      "Chromosome 3 +--------------------------+\n",
      "| AVG(locus.ProteinLength) |\n",
      "+--------------------------+\n",
      "|          252.0           |\n",
      "+--------------------------+\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      "Chromosome 4 +--------------------------+\n",
      "| AVG(locus.ProteinLength) |\n",
      "+--------------------------+\n",
      "|          277.5           |\n",
      "+--------------------------+\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      "Chromosome 5 +--------------------------+\n",
      "| AVG(locus.ProteinLength) |\n",
      "+--------------------------+\n",
      "|    271.2857142857143     |\n",
      "+--------------------------+\n",
      "{1: [(258.75,)], 2: [(215.5,)], 3: [(252.0,)], 4: [(277.5,)], 5: [(271.2857142857143,)]}\n"
     ]
    }
   ],
   "source": [
    "#Average of the protein lenght and creating a dicc to solve the data\n",
    "exe4 = {}\n",
    "for n in range(1,6):\n",
    "    x = \"AT\" + str(n) + \"%\"\n",
    "    exe = %sql SELECT AVG(locus.ProteinLength) FROM germplasm RIGHT JOIN locus ON germplasm.Locus = locus.Locus WHERE germplasm.Locus LIKE :x ;  \n",
    "    exe4[n] = exe\n",
    "    print(\"Chromosome\", n, exe)\n",
    "print(exe4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'exe1' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-144-e66df9489ab1>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mresult_exam2\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mexe1\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mexe2\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mexercise4_exam2\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"exe4.csv\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"w\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mexercise4_exam2\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult_exam2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mexercise4_exam2\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"exe4.csv\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"a\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mexercise4_exam2\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexe3\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'exe1' is not defined"
     ]
    }
   ],
   "source": [
    "result_exam2 = exe1 + exe2\n",
    "exercise4_exam2 = open(\"exe4.csv\", \"w\")\n",
    "exercise4_exam2.write(str(result_exam2))\n",
    "exercise4_exam2 = open(\"exe4.csv\", \"a\")\n",
    "exercise4_exam2.write(exe3)\n",
    "exercise4_exam2.write(exe4)\n",
    "exercise4_exam2.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
